back/sssscs/src/main/java/com/ib/util/DTOModelMapper.java

```diff
@@ -0,0 +1,89 @@
+import org.springframework.web.bind.WebDataBinder;
+import org.springframework.web.bind.support.WebDataBinderFactory;
+import org.springframework.web.context.request.NativeWebRequest;
+import org.springframework.web.method.support.ModelAndViewContainer;
+import org.springframework.web.servlet.mvc.method.annotation.RequestResponseBodyMethodProcessor;
+
+import jakarta.persistence.EntityManager;
+import jakarta.persistence.Id;
+import jakarta.validation.constraints.NotNull;
+import java.io.IOException;
+import java.lang.annotation.Annotation;
+import java.lang.reflect.Field;
+import java.lang.reflect.Type;
+import java.util.Collections;
+
+// Copied from https://auth0.com/blog/automatically-mapping-dto-to-entity-on-spring-boot-apis/
+public class DTOModelMapper extends RequestResponseBodyMethodProcessor {
+    private static final ModelMapper modelMapper = new ModelMapper();
+
+    private EntityManager entityManager;
+
+    public DTOModelMapper(ObjectMapper objectMapper, EntityManager entityManager) {
+        super(Collections.singletonList(new MappingJackson2HttpMessageConverter(objectMapper)));
+        this.entityManager = entityManager;
+    }
+
+    @Override
+    public boolean supportsParameter(MethodParameter parameter) {
+        return parameter.hasParameterAnnotation(DTO.class);
+    }
+
+    @Override
+    protected void validateIfApplicable(WebDataBinder binder, MethodParameter parameter) {
+        binder.validate();
+    }
+
+    @Override
+    public Object resolveArgument(MethodParameter parameter, ModelAndViewContainer mavContainer, NativeWebRequest webRequest, WebDataBinderFactory binderFactory) throws Exception {
+        Object dto = super.resolveArgument(parameter, mavContainer, webRequest, binderFactory);
+        Object id = getEntityId(dto);
+        if (id == null) {
+            return modelMapper.map(dto, parameter.getParameterType());
+        } else {
+            Class entityType = parameter.getParameterType();
+            Object persistedObject = entityManager.find(entityType, id);
+            if (persistedObject == null) {
+                throw new EntityNotFoundException(entityType, (Long) id);
+            }
+            modelMapper.map(dto, persistedObject);
+            return persistedObject;
+        }
+    }
+
+    @Override
+    protected Object readWithMessageConverters(HttpInputMessage inputMessage, MethodParameter parameter, Type targetType) throws IOException, HttpMediaTypeNotSupportedException, HttpMessageNotReadableException {
+        for (Annotation ann : parameter.getParameterAnnotations()) {
+            DTO dtoType = AnnotationUtils.getAnnotation(ann, DTO.class);
+            if (dtoType != null) {
+                return super.readWithMessageConverters(inputMessage, parameter, dtoType.value());
+            }
+        }
+        throw new RuntimeException();
+    }
+
+    private Object getEntityId(@NotNull Object dto) {
+        for (Field field : dto.getClass().getDeclaredFields()) {
```

Comment:
`getDeclaredFields()` could cause problems if we have entities that inherit from other entities (inherited `@Id` field). Will this even be an issue for us; would we even want to have an entity class hierarchy?

Reply:
If it becomes an issue, we could replace it with `getFields()`. Probably just leave it like this for now.